library/clinical_rules.php

```diff
@@ -1042,7 +1045,37 @@ function test_rules_clinic($provider = '', $type = '', $dateTarget = '', $mode =
                     }
 
                     // Check if pass filter
-                    $passFilter = test_filter($rowPatient['pid'], $rowRule['id'], $dateFocus);
+                    /*
+                    HR: changed behavior of test_filter(), and will pass it $dateTarget instead of $dateFocus. Will also call test_filter() only once, not for each $dateFocus value
+                    test_filter() looks for patient data entered prior to $dateTarget timepoint
+
+                    test_filter() was returning
+                    false if an inclusion filter does not succeed
+                    "EXCLUDED" if an exclusion filter succeeds
+                    otherwise true
+
+                    Changed so it now returns:
+                    -- if any required inclusions fail, return false
+                    -- if there are no required inclusions, and some optional inclusions exist, and any optional inclusions succeed,
+                    and either exclusions don't exist or exclusions don't succeed, return true
+                    -- if all inclusions are optional, and none succeed, return false
+                    -- if there are no inclusions, and there are exclusions, and exclusions do not succeed, return true
+                    -- if exclusions succeed (checked only if there are no inclusions, or if inclusions succeed), return 'EXCLUDED'
+                    -- if no inclusions or exclusions, return false
```

Comment:
In the case with no filters, would return `true` to keep the current behavior. A use case is the current `Data Entry - Social Security Number` rule which has no filter thus essentially working on all patients.

Reply:
I checked the social security rule, because I wanted to try one that didn't have any demographic filters and ensure that it would be applied to all patients. 2 issues:
1) The rule was none of: active, passive, or patient. I had to make it one of these three to get it to evaluate when refreshing the reminder widget or reminder details page. Is there a report that would run it without it having one of these three attributes? The rule was also not part of a plan. Adding it to a plan, but without setting one of these three attribute, did not cause the rule to ge evaluated for widget/details page. I guess that was expected.
2) After setting the rule type to "passive", I then tried to unset this checkbox. On saving, I get this error, and the rule detail page won't show:
PHP Fatal error:  Uncaught TypeError: in_array(): Argument #2 ($haystack) must be of type array, string given in C:\xampp\htdocs\openemr\interface\super\rules\library\RuleManager.php:532
Do you want me to try to fix this? (I haven't looked much at this file)

Reply:
Just put in a PR to fix the toggle issue (this is related to more recent PHP versions that force types in the core php functions (the parameter needs to be an empty array rather than an empty string)):
https://github.com/openemr/openemr/pull/7179